File: C#/PBTKata.Tests/Bank/Solution/AccountBuilder.cs
using FsCheck;
using PBTKata.Bank;
using static FsCheck.Arb.Default;

namespace PBTKata.Tests.Bank.Solution
{
    public static class AccountBuilder
    {
        public static Account Empty() => new(0, false, 200m.ToAmountUnsafe(), null);
    }

    public static class AccountExtensions
    {
        private static Amount ArbitraryAmount() =>
            Decimal()
                .Filter(x => x > 0)
                .Generator
                .Sample(1, 1)
                .Head
                .ToAmountUnsafe();

        public static Account WithEnoughMoney(this Account account, Withdraw command) =>
            account with {Balance = (command.Amount + ArbitraryAmount()).Value};

        public static Account WithInsufficientBalance(this Account account, Withdraw command) =>
            account with {Balance = (command.Amount - ArbitraryAmount()).Value};

        public static Account WithdrawAmountReachingMaxWithdrawal(this Account account, Withdraw command) =>
            account with {MaxWithdrawal = command.Amount - ArbitraryAmount()};

        public static Account WithoutReachingMaxWithdrawal(this Account account, Withdraw command) =>
            account with {MaxWithdrawal = command.Amount + ArbitraryAmount()};

        public static Account WithoutOverDraftAuthorized(this Account account) =>
            account with {IsOverdraftAuthorized = false};

        public static Account WithOverDraftAuthorized(this Account account) =>
            account with {IsOverdraftAuthorized = true};
    }
}

File: java/src/test/java/bank/solution/AccountBuilder.java
package bank.solution;

import bank.Account;
import bank.Withdraw;
import io.vavr.collection.List;
import io.vavr.collection.Seq;
import lombok.AllArgsConstructor;
import lombok.With;

import java.util.Random;

@With
@AllArgsConstructor
public class AccountBuilder {
    private static final Random random = new Random();
    private final double balance;
    private final boolean isOverdraftAuthorized;
    private final double maxWithdrawal;
    private final Seq<Withdraw> withdraws;

    public static AccountBuilder newAccount() {
        return new AccountBuilder(0, false, 0, List.empty());
    }

    private static Double arbitraryAmount() {
        return random.nextDouble();
    }

    public AccountBuilder withInsufficientBalance(Withdraw command) {
        return this.withBalance(command.amount().value() - arbitraryAmount());
    }

    public AccountBuilder withdrawAmountReachingMaxWithdrawal(Withdraw command) {
        return this.withMaxWithdrawal(command.amount().value() - arbitraryAmount());
    }

    public AccountBuilder withoutReachingMaxWithdrawal(Withdraw command) {
        return this.withMaxWithdrawal(command.amount().value() + arbitraryAmount());
    }

    public AccountBuilder withoutOverdraft() {
        return this.withOverdraftAuthorized(false);
    }

    public AccountBuilder withOverdraft() {
        return this.withOverdraftAuthorized(true);
    }

    public AccountBuilder withEnoughMoney(Withdraw command) {
        return this.withBalance(command.amount().value() + arbitraryAmount());
    }

    public Account build() {
        return new Account(balance, isOverdraftAuthorized, maxWithdrawal, withdraws);
    }
}
